README.md
# normalizer
Normalizes data extracted with scraper. Normalizer will parse data json files and populate meaningful data into the transformed data json.

## Features
- Parse and download gallery images
- Parse and download thumbnail image
- Unique UUIDv4 image filename generation
- Prevent malicious file downloads through image filetype detection
- Parse phone number from details
- Write transformed data into a separate file
- Map location to ids
- Add default values

## Getting Started
### How to install

`npm install`

### Running in development mode
Development mode executes only the data/example.json file. In this mode, you will also see the maximum level of logs.

`npm start`

### Running in production mode
Production mode scans all the data files inside data/. In this mode, you will only see warning and error logs.

`npm run start-prod`

### Clean up
You may run the following to clean up downloaded files and normalized data files.

`npm run clean`

## Data JSON
Takes raw data JSON scraped from website and outputs normalized data JSON with image files downloaded into attachs directory.

### Raw Data JSON
```json
{
  "title": "法拉盛小单间出租包水电网",
  "category": "法拉盛租房 FlushingA邮编11355，11354。非中介无佣金出租",
  "text1": "",
  "text2": "",
  "text3": "",
  "num1": "",
  "num2": "",
  "select1": "",
  "select2": "",
  "select3": "",
  "select4": "",
  "select5": "",
  "city": "New York",
  "area": "Flushing",
  "subarea": "",
  "date": "2019/7/26",
  "details": "位于法拉盛SANDFORD AVE，超近MAIN ST商业区，5分钟到7号地铁站，附近有数个大超市，全包水电网，适合单身，上班族，留学生，等等。价钱面谈，提供床架与衣柜，少煮食。\r \r 小间单房：单人$400 - 只限年轻人\r \r 有意者电：646-756-0333\r \r 短信：646-756-033",
  "gallery": [
    "http://c.dadi360.com/c/posts/downloadAttach/722638.page"
  ],
  "longitude": "",
  "latitude": "",
  "photo": "",
  "contact_person": "游客",
  "contact_phone": "",
  "contact_qq": "",
  "contact_address": "",
  "depth": 1,
  "url": "http://c.dadi360.com/c/posts/list/2037621.page"
}
```

### Normalized Data JSON
```json
{
  "title": "法拉盛小单间出租包水电网",
  "text1": null,
  "text2": null,
  "text3": null,
  "num1": null,
  "num2": null,
  "select1": null,
  "select2": null,
  "select3": null,
  "select4": null,
  "select5": null,
  "date": 1564124400,
  "details": "位于法拉盛SANDFORD AVE，超近MAIN ST商业区，5分钟到7号地铁站，附近有数个大超市，全包水电网，适合单身，上班族，留学生，等等。价钱面谈，提供床架与衣柜，少煮食。\r \r 小间单房：单人$400 - 只限年轻人\r \r 有意者电：646-756-0333\r \r 短信：646-756-033",
  "gallery": [
    "attachs/2019/07/26/thumb_4665113d-b2c8-4246-99bf-d24be223751e.jpg"
  ],
  "longitude": "",
  "latitude": "",
  "photo": "attachs/2019/07/26/thumb_4665113d-b2c8-4246-99bf-d24be223751e.jpg",
  "contact_person": "游客",
  "contact_phone": "6467560333",
  "contact_qq": "",
  "contact_address": "",
  "depth": 1,
  "url": "http://c.dadi360.com/c/posts/list/2037621.page",
  "cate_id": 47,
  "city_id": 1,
  "area_id": 5,
  "business_id": 1,
  "urgent_date": "2019-07-26",
  "top_date": "2019-07-26",
  "audit": 1,
  "user_id": 2
}
```


CHANGELOG.md
#### 1.5.6 (2019-08-31)

##### Chores

*  Update package ([f4069261](https://github.com/wchen02/normalizer/commit/f4069261e3e584c3e9de715f9305235e37842ffc))

##### Other Changes

*  fix typos in code ([c548519e](https://github.com/wchen02/normalizer/commit/c548519e83fe1f59f8061c527392114a94e7c858))

#### 1.5.5 (2019-08-25)

##### Bug Fixes

*  Fix bug that load and archives files before all data json are normalized ([62ddff9c](https://github.com/wchen02/normalizer/commit/62ddff9cf122541d0230e3a3fa896636713ea309))

#### 1.5.4 (2019-08-20)

##### Other Changes

* **options:**  dropped data env var ([7703fef5](https://github.com/wchen02/normalizer/commit/7703fef560f4997bbb61c325ad4394c8cb7fbff1))

#### 1.5.3 (2019-08-19)

##### Refactors

*  Refactor dir env to work with or without ending '/' ([24fe5b24](https://github.com/wchen02/normalizer/commit/24fe5b2422cfa508b8ccb6a064d46e9696c260fe))

#### 1.5.2 (2019-08-16)

##### Refactors

*  Refactor to remove global variables and move dependencies to the method signature ([963fa87d](https://github.com/wchen02/normalizer/commit/963fa87d7b5cec36dd674747be53fd48ab645448))

#### 1.5.1 (2019-08-16)

##### Bug Fixes

*  add missing await when writing download image into disk ([9ec67eee](https://github.com/wchen02/normalizer/commit/9ec67eeec31c18be39736fc02dc4fed4f2b13dd8))

### 1.5.0 (2019-08-16)

##### Chores

*  remove unused env var ([4002d3fb](https://github.com/wchen02/normalizer/commit/4002d3fb8983bb1da0905b1eac24e2a733af5f03))

##### New Features

*  create download dir and normalized dir if not exist ([707a38ab](https://github.com/wchen02/normalizer/commit/707a38abceca8e612f4a14439998348c9099c16c))

#### 1.4.1 (2019-08-16)

##### Bug Fixes

*  long attach file name causing loader to not able to safe to db ([e45801be](https://github.com/wchen02/normalizer/commit/e45801be983c1eb5a1fbed43ec20287ca6ed6744))

### 1.4.0 (2019-08-16)

##### New Features

*  Add getOptions method into lib ([642c55d7](https://github.com/wchen02/normalizer/commit/642c55d7e3e95682a94bd96ef2b95ff38225ae39))

### 1.3.0 (2019-08-16)

##### Chores

*  update example .env ([4a48af18](https://github.com/wchen02/normalizer/commit/4a48af18fda1af9e3edb189dba48610cab4acb94))

##### New Features

*  Make directories configurable vai env var ([1c674eac](https://github.com/wchen02/normalizer/commit/1c674eacac02f185d66cc2c922e0ba6986a0de17))

### 1.2.0 (2019-08-11)

##### New Features

*  download and generate thumbnail ([95d1b627](https://github.com/wchen02/normalizer/commit/95d1b6272c461470d6fa4039164122d8d698d596))

### 1.1.0 (2019-08-03)

##### New Features

*  Support processing data file with array of dataJson, added max concurrency configuration, normalized file are now written to unique key ([0b13e695](https://github.com/wchen02/normalizer/commit/0b13e695c2f4560f3ddc2a73a767a1a5d5fe36d7))
*  Support pass in data file to process in development via env variable ([3ca76fae](https://github.com/wchen02/normalizer/commit/3ca76fae7a4bb2b78adefd0793941300b47c9611))

##### Bug Fixes

*  Forgot to convert last time ([c8800c7b](https://github.com/wchen02/normalizer/commit/c8800c7be519e5ac834779dc5a01cf95374739fa))

## 1.0.0 (2019-07-27)

##### Documentation Changes

*  Prepare Readme for major release ([0717abac](https://github.com/wchen02/normalizer/commit/0717abac6a17393ef2a7a398386af77f09d9d470))
*  Update Readme ([669a93d0](https://github.com/wchen02/normalizer/commit/669a93d0b1b4e4284af47381d2438068ebc31dd2))

##### Refactors

*  Move env vars from script to dotenv, updated example json ([ad3b236b](https://github.com/wchen02/normalizer/commit/ad3b236b9d636ca3a792d506c87346f1f1ce1343))

### 0.6.0 (2019-07-27)

##### New Features

*  Rename download dir to attachs, download thumbnail image, add error handling, standardized logging ([696caf76](https://github.com/wchen02/normalizer/commit/696caf76baea76a9fc43b091e9c0dd642482fdeb))
*  Add more default values ([81c91836](https://github.com/wchen02/normalizer/commit/81c91836af2812a57dc5245d93361b0e2c08965f))
*  Add missing field default values ([41cb962d](https://github.com/wchen02/normalizer/commit/41cb962d8a8cb3beb4c5e383f9c1c67d3eed77a7))

### 0.5.0 (2019-07-27)

##### Breaking Changes

*  update data json schema ([93270ad8](https://github.com/wchen02/normalizer/commit/93270ad828f0fd1d0821f70440e7d4c223c748bb))

##### Chores

*  keeping normalized dir, remove example.json ([4de43212](https://github.com/wchen02/normalizer/commit/4de43212d91989b52a869f649d71a73139d5171c))

##### New Features

*  Parse and convert date to timestamp ([603fde0a](https://github.com/wchen02/normalizer/commit/603fde0a710b6637f31ef1eb9e95f91b91180822))
*  map city, area, and subarea to ids ([55264aab](https://github.com/wchen02/normalizer/commit/55264aabe4fa024849a95c934a2b26a2e7a861fb))
*  map category to category Id ([a761c5b7](https://github.com/wchen02/normalizer/commit/a761c5b795bca5fe30fd00062ebc9dc89296adee))

##### Bug Fixes

*  update to use the full category name ([3a288315](https://github.com/wchen02/normalizer/commit/3a2883150d5c34a36c0d5fc1408e28eba93b5525))

#### 0.4.1 (2019-07-24)

##### Documentation Changes

*  Update readme ([42d110ab](https://github.com/wchen02/normalizer/commit/42d110abfbccc223744d3c7b638f29b10141aa95))

##### Bug Fixes

*  Fix wrong log level in start-prod script ([03ab3f96](https://github.com/wchen02/normalizer/commit/03ab3f96131a85169a1b243d2b8afeaf2bd45cf1))

### 0.4.0 (2019-07-24)

##### Documentation Changes

*  Update readme ([c2f2a3d6](https://github.com/wchen02/normalizer/commit/c2f2a3d6ec1960a459490b1cac181b2e163a4b5f))

##### New Features

*  Write transformed file to data dir, update clean script to clean normalized data dir as well ([e1ac48eb](https://github.com/wchen02/normalizer/commit/e1ac48eb264ac85941f16dfc6dabd420c382fd81))

##### Performance Improvements

*  avoid scanning data dir for development mode ([109749ea](https://github.com/wchen02/normalizer/commit/109749eaeacca188b6e45919ebaca6c4fe5bdc63))


index.js
const { findNumbers } = require('libphonenumber-js');
const fs = require('fs');
const fileType = require('file-type');
const log = require('loglevel');
const { promisify } = require('util');
const download = require('download');
const jsonfile = require('jsonfile');
const mkdirp = require('mkdirp');
const uuid = require('uuid/v4');
const { format, getTime } = require('date-fns');
const async = require('async');
const thumb = require('node-thumbnail').thumb;

async function openFile(filename) {
    log.info(`Opening ${ filename }`)
    let dataJson;
    try {
        dataJson = await jsonfile.readFile(filename);
    } catch (err) {
        log.error(`Error opening file: "${ filename }"`);
        log.error(err);
    }
    log.debug(dataJson);
    return dataJson;
}

async function writeFile(filename, dataJson) {
    log.info(`Writing Transformed Data JSON to ${ filename }`)
    log.debug(dataJson);
    try {
        dataJson = await jsonfile.writeFile(filename, dataJson, { spaces: 2 });
    } catch (err) {
        log.error(`Error writing to file: "${ filename }"`);
        log.error(err);
    }
    log.debug('Done');
    return dataJson;
}

async function downloadImage(dataJson, imgUrl, options) {
    if (!imgUrl) {
        return;
    }

    const date = new Date(dataJson.date);
    const year = format(date, 'YYYY');
    const month = format(date, 'MM');
    const day = format(date, 'DD');

    const attachsDir = options.dir.download + '/' + year + '/' + month + '/' + day + '/';
    makeDirectories(attachsDir);
    
    log.info(`Download image ${ imgUrl }`);
    let fileData;
    try {
        fileData = await download(imgUrl);
    } catch (err) {
        log.error(`Error downloading image ${ imgUrl }`);
        log.error(err);
    }

    const fileDataType = fileType(fileData);
    if (!fileDataType || !fileDataType.mime.includes('image/')) {
        log.warn('Invalid filetype, skipping.');
        return;
    }

    const writeFileAsync = promisify(fs.writeFile);
    const strippedUuid = uuid().replace(/-/g, '');
    const filename = attachsDir + strippedUuid + '.' + fileDataType.ext;
    try {
        await writeFileAsync(filename, fileData);
    } catch (err) {
        log.error(`Error writing file ${ filename }`);
        log.error(err);
    }
    log.debug(`Downloaded to ${ filename }`);

    return filename;
}

async function generateThumbnailImage(imgFilename, width, height) {
    if (!imgFilename) {
        return;
    }

    const indexOfLastDir = imgFilename.lastIndexOf('/') + 1;
    const imageFileDir = imgFilename.substring(0, indexOfLastDir);
    const imgName = imgFilename.substring(indexOfLastDir);
    const thumbnailFilename = imageFileDir + 'thumb_' + imgName;

    log.info(`Generating ${ width }x${ height } thumbnail from ${ imgFilename }`);
    try {
        await thumb({
            source: imgFilename,
            prefix: 'thumb_',
            suffix: '',
            destination: imageFileDir,
            width,
            logger: function(message) {
                log.debug(message);
            }
        });
    } catch(err) {
        log.error(`Error generating thumbnail from ${ imgFilename }`);
        log.error(err);        
    }
    log.debug(`Generated thumbnail ${ thumbnailFilename }`);

    return thumbnailFilename;
}

async function downloadThumbnailImage(dataJson, transformedDataJson, options) {
    let thumbnailFilename;

    if (dataJson.photo) {
        const imgFilename = await downloadImage(dataJson, dataJson.photo, options);
        thumbnailFilename = await generateThumbnailImage(imgFilename, options.lifeListThumbnail.width, options.lifeListThumbnail.height);
    } else if (transformedDataJson.gallery.length) {
        thumbnailFilename = transformedDataJson.gallery[0];
    }
    transformedDataJson.photo = thumbnailFilename;
}

async function downloadGalleryImages(dataJson, transformedDataJson, options) {
    if (!dataJson.gallery.length) {
        return;
    }

    log.info(`Found ${ dataJson.gallery.length } gallery images`);
    const newGalleryPaths = await Promise.all(dataJson.gallery.map(async (imgUrl) => {
        return downloadImage(dataJson, imgUrl, options);
    }));
    log.debug('Gallery images: ');
    log.debug(newGalleryPaths);

    const newGalleryThumbnailPaths = await Promise.all(newGalleryPaths.map(async (imgFilename) => {
        return generateThumbnailImage(imgFilename, options.lifeDetailThumbnail.width, options.lifeDetailThumbnail.height);
    }));

    transformedDataJson.gallery = newGalleryThumbnailPaths;
}

async function parsePhoneNumber(dataJson, transformedDataJson) {
    if (!dataJson.contact_phone) {
        log.info('Parsing phone number');
        const phoneNumber = findNumbers(dataJson.details, 'US', { v2: true });
        if (phoneNumber.length) {
            log.debug(`Found ${ phoneNumber[0].number.nationalNumber }`);
            transformedDataJson.contact_phone = phoneNumber[0].number.nationalNumber;
        }
    }
}

function mapCategoryToId(dataJson, transformedDataJson) {
    const categoryIdMap = {
        '纽约皇后区·房屋出租': 47,
        '纽约·艾姆赫斯特·房屋出租(Elmhurst & Corona)': 47,
        '纽约·艾姆赫斯特·房屋出租(Elmhurst & Corona)': 47,
        '法拉盛租房 FlushingA邮编11355，11354。非中介无佣金出租': 47,
        '租房(布碌仑)': 47,
        '租房(曼哈顿)': 47,
        '长岛·新泽西·其它区': 47,
        '房屋买卖': 49,
        '商铺租让，办公楼租让，生意租让': 51,
        '商铺租让，办公楼租让，生意租让': 51,
        '顺风车(Car Pool)': 46,
        '家庭旅馆•床位出租•机票•旅行社•纽约华人旅店': 104,
        '大小搬家·纽约搬家公司·法拉盛搬家公司': 108,
        '电召车·机场接送·购物旅游·旅游包车·旅游租车·巴士': 103,
        '纽约社区信息 纽约社区活动 纽约周末活动': 44,
        '教会·宗教': 44,
        '其它': 147,
        '大地问答': 44,
        '房东房客': 47,
        '大地聊天室': 44,
        '好玩網頁遊戲·娱乐': 44,
        '招聘·请员工': 89,
        '甲店·美容Spa·发廊·纹身店请人': 83,
        '餐馆请人·KTV请人·奶茶店请人·水吧·酒吧请人': 74,
        '推拿按摩请人·请前台·请司机': 69,
        '求职': 38,
        '销售·兼职·代购招聘·合作伙伴': 68,
        '网络公司招聘·网络电商兼职': 84,
        '二手买卖': 19,
        '手机买卖·手机维修': 1,
        '汽车买卖': 21,
        '电脑买卖': 14,
        '货品买卖': 19,
        '交友·征婚': 42,
        '生活服务·摄影服务·护照像': 129,
        '装修服务': 95,
        '冷暖水电': 95,
        '防盗报警·警钟·远程近程监控': 129,
        '验屋服务': 129,
        '纽约中餐馆·小吃--华埠，中城': 125,
        '纽约中餐馆·小吃--皇后区，法拉盛': 125,
        '纽约中餐馆·小吃 -- 布碌仑 (Brooklyn)': 125,
        '各类律师': 122,
        '会计•报税': 97,
        '各类保险': 129,
        '贷款·财务·投资·信用卡': 91,
        '入籍•福利': 122,
        '留学•转学•签证•移民': 122,
        '翻译服务·各类公证·认证·委托': 126,
        '电脑·网络·硬件·附件': 98,
        '网页设计·软件开发': 118,
        '印刷•招牌': 112,
        '职业技术培训·才艺学校·英语课程·中文教学': 59,
        '各科补习·补习班·SAT·ACT·升学补习': 58,
        '幼儿园·家庭托儿': 65,
        '美容•发廊•Spa': 116,
        '推拿·按摩·理疗·痛症治疗': 116,
        '各科医生·理疗·保健': 128,
        '考牌练车': 109,
        '快递•货运': 99,
        '免费广告(加拿大)': 147,
    };

    if (dataJson.category && categoryIdMap[dataJson.category]) {
        transformedDataJson.cate_id = categoryIdMap[dataJson.category];
    } else {
        transformedDataJson.cate_id = 147
    }

    log.info(`Mapped category ${ dataJson.category } to ${ transformedDataJson.cate_id }`);
    delete transformedDataJson.category;
}

function mapCityToId(dataJson, transformedDataJson) {
    const cityIdMap = {
        'New York': 1
    }
    if (dataJson.city && cityIdMap[dataJson.city]) {
        transformedDataJson.city_id = cityIdMap[dataJson.city];
    } else {
        transformedDataJson.city_id = 1;
    }

    log.info(`Mapped city ${ dataJson.city } to ${ transformedDataJson.city_id }`);
    delete transformedDataJson.city;
}

function mapAreaToId(dataJson, transformedDataJson) {
    const areaIdMap = {
        'Queens': 1,
        'Flushing': 5,
        'Elmhurst': 6,
        'Brooklyn': 7,
        'Manhattan': 8,
        'Bronx': 9,
        'Staten Island': 10,
        'Other': 11,
        'Long Island': 12,
    }
    if (dataJson.area && areaIdMap[dataJson.area]) {
        transformedDataJson.area_id = areaIdMap[dataJson.area];
    } else {
        transformedDataJson.area_id = 1;
    }

    log.info(`Mapped area ${ dataJson.area } to ${ transformedDataJson.area_id }`);
    delete transformedDataJson.area;
}

function mapSubareaToBusinessId(dataJson, transformedDataJson) {
    const subareaToBusinessIdMap = {
        'Flushing': 1
    }
    if (dataJson.subarea && subareaToBusinessIdMap[dataJson.subarea]) {
        transformedDataJson.business_id = subareaToBusinessIdMap[dataJson.subarea];
    } else {
        transformedDataJson.business_id = 1;
    }

    log.info(`Mapped subarea ${ dataJson.subarea } to ${ transformedDataJson.business_id }`);
    delete transformedDataJson.subarea;
}

function convertDateToUnixTimestamp(dataJson, transformedDataJson) {
    if (!dataJson.date) {
        return;
    }
    const date = new Date(dataJson.date);
    const unixTimestamp = getTime(date)/1000;

    transformedDataJson.create_time = unixTimestamp;
    transformedDataJson.last_time = unixTimestamp;    
    log.info(`Converted ${ dataJson.date } to ${ unixTimestamp }`);

    delete transformedDataJson.date;
}

function addMissingDefaultFields(dataJson, transformedDataJson) {
    const date = dataJson.date ? new Date(dataJson.date) : new Date();
    const formattedDate = format(date, 'YYYY-MM-DD');

    if (!dataJson.urgent_date) {
        transformedDataJson.urgent_date = formattedDate;
    }
    if (!dataJson.top_date) {
        transformedDataJson.top_date = formattedDate;
    }
    if (!dataJson.text1) {
        transformedDataJson.text1 = null;
    }
    if (!dataJson.text2) {
        transformedDataJson.text2 = null;
    }
    if (!dataJson.text3) {
        transformedDataJson.text3 = null;
    }
    if (!dataJson.num1) {
        transformedDataJson.num1 = null;
    }
    if (!dataJson.num2) {
        transformedDataJson.num2 = null;
    }
    if (!dataJson.select1) {
        transformedDataJson.select1 = null;
    }
    if (!dataJson.select2) {
        transformedDataJson.select2 = null;
    }
    if (!dataJson.select3) {
        transformedDataJson.select3 = null;
    }
    if (!dataJson.select4) {
        transformedDataJson.select4 = null;
    }
    if (!dataJson.select5) {
        transformedDataJson.select5 = null;
    }
    if (!dataJson.audit) {
        transformedDataJson.audit = 1;
    }
    if (!dataJson.user_id) {
        transformedDataJson.user_id = 2;
    }
}

function getUniqueKey(dataJson) {
    return dataJson.url.toLowerCase().replace(/\W/g, '_');
}

async function processDataJson(dataJson, options) {
    try {
        // copied a new object out of dataJson
        const transformedDataJson = JSON.parse(JSON.stringify(dataJson));

        await Promise.all([
            downloadGalleryImages(dataJson, transformedDataJson, options),
            parsePhoneNumber(dataJson, transformedDataJson),
            mapCategoryToId(dataJson, transformedDataJson),
            mapCityToId(dataJson, transformedDataJson),
            mapAreaToId(dataJson, transformedDataJson),
            mapSubareaToBusinessId(dataJson, transformedDataJson),
            addMissingDefaultFields(dataJson, transformedDataJson),
        ]);
        await downloadThumbnailImage(dataJson, transformedDataJson, options);
        await convertDateToUnixTimestamp(dataJson, transformedDataJson);

        const uniqueKey = await getUniqueKey(dataJson);
        await writeFile(options.dir.normalizedData + '/' + uniqueKey + '.json', transformedDataJson);
    } catch (err) {
        log.error('Error processing dataJson');
        log.error(dataJson);
        log.error(err);
    }

}
async function processFile(filename, options) {
    try {
        const dataJson = await openFile(options.dir.rawData + '/' + filename);
        if (Array.isArray(dataJson)) {            
            await async.mapLimit(dataJson, options.maxConcurrency, async (json) => {
                await processDataJson(json, options);
            });
        } else {
            await processDataJson(dataJson, options);
        }
    } catch (err) {
        log.error(`Error processing file ${ filename }`);
        log.error(err);
    }
}

function makeDirectories(dir) {
    log.info(`Making directory ${ dir }`);
    try {
        mkdirp.sync(dir);
    } catch (err) {
        log.error(`Error to make dir ${ dir }`);
        log.error(err);
    }

    return dir;
}

async function run(options) {
    const { 
        logLevel, 
        isDevelopment,
        dataFile,
        dir,
    } = options;

    log.setLevel(logLevel);

    makeDirectories(dir.download);
    makeDirectories(dir.normalizedData);
    
    if (isDevelopment) {
        await processFile(dataFile, options);
    } else {
        const readdirAsync = promisify(fs.readdir)

        let files;
    
        try {
            log.info(`Scanning data directory ${ dir.rawData }`);
            files = await readdirAsync(dir.rawData);
            log.info(`Found ${ files.length } data files.`);
        } catch (err) {
            log.error(`Error scanning directory ${ dir.rawData }`);
            log.error(err);
            return;
        }

        await Promise.all(files.map(filename => {
            processFile(filename, options);
        }));
    }
}

function getOptions(options) {
    return {
        maxConcurrency: options.MAX_CONCURRENCY,
        logLevel: options.LOG_LEVEL,
        lifeListThumbnail: {
            width: options.LIFE_LIST_THUMBNAIL_WIDTH,
            height: options.LIFE_LIST_THUMBNAIL_HEIGHT,
        },
        lifeDetailThumbnail: {
            width: options.LIFE_DETAIL_THUMBNAIL_WIDTH,
            height: options.LIFE_DETAIL_THUMBNAIL_HEIGHT,
        },
        isDevelopment: options.DEVELOPMENT,
        dataFile: options.DATA_FILE,
        dir: {
            download: options.DOWNLOAD_DIR,
            rawData: options.RAW_DATA_DIR,
            normalizedData: options.NORMALIZED_DATA_DIR,
        }
    }
}

module.exports = {
    run,
    getOptions,
};
